alf/algorithms/sac_algorithm.py

```diff
@@ -0,0 +1,365 @@
+            critic1, critic1_state = self._critic_network1(
+                critic_input,
+                step_type=exp.step_type,
+                network_state=state.critic1)
+
+            critic2, critic2_state = self._critic_network2(
+                critic_input,
+                step_type=exp.step_type,
+                network_state=state.critic2)
+
+            target_q_value = tf.minimum(critic1, critic2)
+
+        dqda = tape.gradient(target_q_value, action)
+
+        def actor_loss_fn(dqda, action):
+            if self._dqda_clipping:
+                dqda = tf.clip_by_value(
+                    dqda, -self._dqda_clipping,
+                    self._dqda_clipping)
+            loss = losses.element_wise_squared_loss(
+                tf.stop_gradient(dqda + action), action)
+            loss = tf.reduce_sum(loss, axis=loss.shape[1:])
+            return loss
+
+        actor_loss = tf.nest.map_structure(actor_loss_fn, dqda, action)
+        actor_loss += tf.stop_gradient(tf.exp(self._log_alpha)) * log_pi
+
+        state = SacActorState(
+            actor=actor_state,
+            critic1=critic1_state,
+            critic2=critic2_state)
+        info = SacActorInfo(
+            log_pi=log_pi, critic1=critic1,
+            critic2=critic2, loss=LossInfo(loss=actor_loss, extra=actor_loss))
+        return state, info
+
+    def _critic_train_step(self, exp: Experience, state: SacCriticState):
+        action, log_pi, actor_state = self._actions_and_log_probs(exp, state)
+
+        critic_input = (exp.observation, exp.action)
+        critic1, critic1_state = self._critic_network1(
+            critic_input,
+            step_type=exp.step_type,
+            network_state=state.critic1)
+
+        critic2, critic2_state = self._critic_network2(
+            critic_input,
+            step_type=exp.step_type,
+            network_state=state.critic2)
+
+        target_critic_input = (exp.observation, action)
+
+        target_critic1, target_critic1_state = self._target_critic_network1(
+            target_critic_input,
+            step_type=exp.step_type,
+            network_state=state.target_critic1)
+
+        target_critic2, target_critic2_state = self._target_critic_network2(
+            target_critic_input,
+            step_type=exp.step_type,
+            network_state=state.target_critic2)
+
+        state = SacCriticState(
+            actor=actor_state,
+            critic1=critic1_state,
+            critic2=critic2_state,
+            target_critic1=target_critic1_state,
+            target_critic2=target_critic2_state)
+
+        info = SacCriticInfo(
+            log_pi=log_pi,
+            critic1=critic1, critic2=critic2,
+            target_critic1=target_critic1,
+            target_critic2=target_critic2)
+
+        return state, info
+
+    def _alpha_train_step(self, exp: Experience, state: SacActorState):
+        action, log_pi, actor_state = self._actions_and_log_probs(exp, state)
```

Comment:
Can we avoid the duplicated computation of self._actions_and_log_probs(exp, state) as _critic_train_step, _actor_train_step all use it.

Reply:
add SacShareState for store actor state for actor, critic, alpha and calc _actions_and_log_probs only once ,but it seems not as clear as before